### DISCUSSION TITLE
Cannot converge after fine-tuning the ASR model.

### DISCUSSION THREAD

**Discussion by Minxiangliu** [NVIDIA-NeMo/NeMo#2759] [Q&A] [answered]
I use this [citrinet_512.yaml](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/citrinet/citrinet_512.yaml) and prepare label like [this](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/quartznet/quartznet_15x5_zh.yaml), and then use [stt_zh_citrinet_512](https://ngc.nvidia.com/catalog/models/nvidia:nemo:stt_zh_citrinet_512) to fine-tuning the model.
The training data is about 50 hours with `train-clean-100.tar.gz` in [LibriSpeech](https://www.openslr.org/12), and the [Chinese data](https://commonvoice.mozilla.org/zh-TW/datasets) is about 50 hours.
But it did not converge normally during training. What could be the reason for this?

![image](https://user-images.githubusercontent.com/30144428/131685569-99bbb46c-c90d-422a-96e8-e4dbead09549.png)

Thank you for your time!

**Comment by titu1994** [comment, marked as answer]
You should directly fine-tune the stt_zh_citrinet_512 model by following the steps in the CTC Finetuning tutorial. That checkpoint should use CER by default as well which might be why you "word error rate" is high - mandarin required CER not WER

**Reply by titu1994** [reply]
Also make sure that the English tokens in the manifest correspond to to the English tokens in the vocabulary provoded in the model, otherwise text will be incorrect for English characters.

**Reply by Minxiangliu** [reply]
I used this [script](https://github.com/NVIDIA/NeMo/blob/stable/scripts/dataset_processing/get_librispeech_data.py) to download the English data. I didn’t change any data in the manifest. I only reduced the amount of data to the time I wanted.
Like this:
```
{"audio_filepath": "/voiceDataSet/LibriSpeech/train-clean-100-processed/32-21634-0000.wav", "duration": 15.58, "text": "there was great rejoicing in the little village in the pond when brownie beaver returned with the good news that there would be no more hunting and fishing and when old grandaddy beaver said that everybody ought to take a holiday to celebrate the occasion"}
{"audio_filepath": "/voiceDataSet/LibriSpeech/train-clean-100-processed/32-21634-0001.wav", "duration": 12.91, "text": "all the villagers said it was a fine idea so they stopped working for once and began to plan the celebration they thought that there ought to be swimming races and tree felling contests"}
{"audio_filepath": "/voiceDataSet/LibriSpeech/train-clean-100-processed/32-21634-0002.wav", "duration": 12.78, "text": "and brownie beaver said that after the holiday was over he would suggest that someone be chosen to go down and thank farmer green for putting the notice on the tree the whole village agreed to brownie's proposal"}
....
```

And the `yaml` file I prepared is as follow below. if I did not prepare the correct English vocabulary, is there a mistake in my way?

```
 labels: &labels [' ', '''', a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r,
                   s, t, u, v, w, x, y, z, 㶧, 䶮, 一, 丁, 七, 萬, 丈, 三, 上, 下, 不, 與, 丐, 醜, 專, 且, 丕,
                   世, 丘, 丙, 業, 叢, 東, 絲, 丞, 丟, 兩, 嚴, 喪, 個, 丫, 中, 豐, 串, 臨, 丸, 丹, 為, 主, 麗, 舉, 乃,
                   久, 么, 义, 之, 乌, 乍, 乎, 乏, 乐, 乒, 乓, 乔, 乖, 乘, 乙, 九, 乞, 也, 习, 乡, 书, 买, 乱, 乳, 乾,
                   了, 予, 爭, 事, 二, 於, 虧, 雲, 互, 五, 井, 亙, 亞, 些, 亟, 亡, 亢, 交, 亥, 亦, 產, 亨, 畝, 享, 京,
                   亭, 亮, 親, 亳, 褻, 人, 億, 什, 仁, 仄, 僅, 僕, 仇, 今, 介, 仍, 從, 崙, 倉, 仔, 仕, 他, 仗, 付, 仙,
                   仞, 仟, 仡, 代, 令, 以, 仨, 儀, 們, 仰, 仲, 件, 價, 任, 份, 仿, 企, 伉, 伊, 伍, 伎, 伏, 伐, 休, 眾,
                   優, 夥, 會, 傘, 偉, 傳, 伢, 傷, 倫, 偽, 佇, 伯, 估, 伴, 伶, 伸, 伺, 似, 伽, 佃, 但, 位, 低, 住, 佐,
.....
```
I will try `CTC Finetuning tutorial` later.

**Reply by titu1994** [reply]
You're changing tokenizer - that is equivalent to training with just good initialization.

The hyper parameters you are using are for from scratch training - LR of 0.05 is too high for 50 hours of eng + mandarin data trained for a few epochs. Your train loss shows this. Further, your dev loss is going down steadily but model does not predict anything - this is due to too short training + too little data. Also ensure you are using CER with model.use_cer: true in config. Wer is always > 100% in Mandarin since mandarin doesn't have spaces.

**Reply by Minxiangliu** [reply]
@titu1994 Hi,
I have trained step by step according to [this tutorials (Sub-word Encoding CTC Model)](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/ASR_CTC_Language_Finetuning.ipynb), in which I used the 50 hours of eng + mandarin data I prepared to execute [this](https://github.com/NVIDIA/NeMo/blob/main/scripts/tokenizers/process_asr_text_tokenizer.py)`(spe_character_coverage=0.98)`. Finally, other settings remain unchanged, including `freeze_encoder=True`, `use_cer=True`, etc.

The results:
![image](https://user-images.githubusercontent.com/30144428/133775522-11f1cb1f-0d9f-447b-9d79-415d42d731e5.png)

So, do I need to prepare more data set next to make the model more accurate?

**Reply by titu1994** [reply]
Nice! CER is going down nicely now. Training isn't finished yet so let it complete it should improve more towards the end of training. At this point yes, you would need more data to improve performance. Also since it's mixed eng + mandarin model, you need to try to see which initialization is a good one - you could start with a mandarin initialized model or an English initialized model too.

**Reply by Minxiangliu** [reply]
@titu1994 Hi,
My training ended at 500 epoch, but I think the model can still converge.

The results:
![image](https://user-images.githubusercontent.com/30144428/133869539-d2afd94f-4f0b-4bab-9a18-8181df4ac115.png)


The questions:
1. Can I continue to fine-tune my final result of model? If so, how much can I suggest for LR or other settings?

2. How many epochs do you recommend for fine-tuning training for this new language? From my example, my 500 epoch is still not enough.

Many thank!

**Reply by titu1994** [reply]
Depends on dataset size. We train LS for 100-1000 epochs and see near 1-2% WER improvement when switching from 100 to 1000 epochs. 

Yes you can continue finetuning, had the max lr as slightly higher than the final LR and continue training.

**Reply by Minxiangliu** [reply]
@titu1994 Hi,
Do you mean that the final LR before the model finishes training is `1.692e-9`, and the LR I fine-tuned must be slightly higher by `1.692e-9`, such as `1.692e-8`? Or you can suggest me if it is the current LR, how much I can adjust to allow me to continue to fine-tune the model.
The result:
![image](https://user-images.githubusercontent.com/30144428/133870142-d4db6154-7e30-497b-9a2e-a0e475666159.png)

Thank you for your help!

**Reply by titu1994** [reply]
You should set a min lr of 1e-5 or 1e-6. This is too small LR.

**Reply by Minxiangliu** [reply]
OK, My model training experience is not enough, thank you for your opinion, I will try.
